fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/JoinOnPredicateCommutativityRule.java

```diff
@@ -0,0 +1,144 @@
+        OptExpression leftChild = input.getInputs().get(0);
+        LogicalJoinOperator leftJoinOp = (LogicalJoinOperator) leftChild.getOp();
+
+        Map<Integer, List<ColumnRefOperator>> childEqColRefMap = getChildEqColRefMap(leftJoinOp);
+
+        LogicalJoinOperator joinOp = (LogicalJoinOperator) input.getOp();
+        List<BinaryPredicateOperator> joinEqOps = Utils.extractConjuncts(joinOp.getOnPredicate())
+                .stream()
+                .filter(JoinPredicateUtils::isEqualBinaryPredicate)
+                .map(ScalarOperator::<BinaryPredicateOperator>cast)
+                .collect(Collectors.toList());
+        for (BinaryPredicateOperator joinEqOp : joinEqOps) {
+            ScalarOperator left = joinEqOp.getChild(0);
+            ScalarOperator right = joinEqOp.getChild(1);
+            if (!(left instanceof ColumnRefOperator) || !(right instanceof ColumnRefOperator)) {
+                continue;
+            }
+            ColumnRefOperator leftColRef = left.cast();
+            ColumnRefOperator rightColRef = right.cast();
+
+            // Key is the col id of replaceable column
+            // Value is the saved column
+            Map<Integer, ColumnRefOperator> replaceableColIds = Maps.newHashMap();
+            if (childEqColRefMap.containsKey(leftColRef.getId())) {
+                replaceableColIds.put(leftColRef.getId(), rightColRef);
+            }
+            if (childEqColRefMap.containsKey(rightColRef.getId())) {
+                replaceableColIds.put(rightColRef.getId(), leftColRef);
+            }
+
+            if (replaceableColIds.isEmpty()) {
+                continue;
+            }
+
+            List<ScalarOperator> otherOps = ListUtils.subtract(Utils.extractConjuncts(joinOp.getOnPredicate()),
+                    Collections.singletonList(joinEqOp));
+
+            for (Map.Entry<Integer, ColumnRefOperator> pair : replaceableColIds.entrySet()) {
+                for (ColumnRefOperator alternativeColRef : childEqColRefMap.get(pair.getKey())) {
+                    BinaryPredicateOperator newEqOp =
+                            new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.EQ, alternativeColRef,
+                                    pair.getValue());
+                    ScalarOperator newJoinOnPredicate = Utils.compoundAnd(
+                            ListUtils.union(Collections.singletonList(newEqOp), otherOps)
+                    );
+                    OptExpression newJoin = OptExpression.create(new LogicalJoinOperator.Builder().withOperator(joinOp)
+                                    .setOnPredicate(newJoinOnPredicate)
+                                    .build(),
+                            input.getInputs());
+                    result.add(newJoin);
+                }
+            }
+        }
+
+        return result;
+    }
+
+    private Map<Integer, List<ColumnRefOperator>> getChildEqColRefMap(LogicalJoinOperator leftJoinOp) {
```

Comment:
maybe you could reuse `ScalarEquivalenceExtractor`

Reply:
Removed